File: test/verification.cpp
#include <memory>
#include <optional>
#include <string>

#include <gtest/gtest.h>

#include "behavior.hpp"
#include "cost_arbitrator.hpp"
#include "priority_arbitrator.hpp"

#include "cost_estimator.hpp"
#include "dummy_types.hpp"


using namespace arbitration_graphs;
using namespace arbitration_graphs_tests;

using DummyPlaceboVerifier = verification::PlaceboVerifier<DummyEnvironmentModel, DummyCommand>;

struct DummyVerifier : public verification::Verifier<DummyEnvironmentModel, DummyCommand> {
    verification::Result::Ptr analyze(const Time& /*time*/,
                                      const DummyEnvironmentModel& /*environmentModel*/,
                                      const DummyCommand& command) const override {
        if (command == wrong_) {
            return std::make_shared<DummyResult>(false);
        }
        return std::make_shared<DummyResult>(true);
    };
    std::string wrong_{"MidPriority"};
};

class CommandVerificationTest : public ::testing::Test {
protected:
    DummyBehavior::Ptr testBehaviorHighPriority = std::make_shared<DummyBehavior>(false, false, "HighPriority");
    DummyBehavior::Ptr testBehaviorMidPriority = std::make_shared<DummyBehavior>(true, false, "MidPriority");
    DummyBehavior::Ptr testBehaviorLowPriority = std::make_shared<DummyBehavior>(true, true, "LowPriority");

    DummyEnvironmentModel environmentModel;

    Time time{Clock::now()};
};


// The default verifyier should be the PlaceboVerifier
TEST_F(CommandVerificationTest, DefaultVerifier) {
    using OptionFlags = PriorityArbitrator<DummyEnvironmentModel, DummyCommand>::Option::Flags;

    PriorityArbitrator<DummyEnvironmentModel, DummyCommand> testPriorityArbitrator;

    testPriorityArbitrator.addOption(testBehaviorHighPriority, OptionFlags::NO_FLAGS);
    testPriorityArbitrator.addOption(testBehaviorHighPriority, OptionFlags::NO_FLAGS);
    testPriorityArbitrator.addOption(testBehaviorMidPriority, OptionFlags::NO_FLAGS);
    testPriorityArbitrator.addOption(testBehaviorLowPriority, OptionFlags::NO_FLAGS);

    ASSERT_TRUE(testPriorityArbitrator.checkInvocationCondition(time, environmentModel));

    testPriorityArbitrator.gainControl(time, environmentModel);

    EXPECT_EQ("MidPriority", testPriorityArbitrator.getCommand(time, environmentModel));
    EXPECT_FALSE(testPriorityArbitrator.options().at(0)->verificationResult_.cached(time));
    EXPECT_FALSE(testPriorityArbitrator.options().at(1)->verificationResult_.cached(time));
    ASSERT_TRUE(testPriorityArbitrator.options().at(2)->verificationResult_.cached(time));
    // LowPriority could have been verified or not, so don't test it here

    EXPECT_TRUE(testPriorityArbitrator.options().at(2)->verificationResult_.cached(time).value()->isOk());
}


// Same as for DefaultVerification
TEST_F(CommandVerificationTest, PlaceboVerifier) {
    using OptionFlags = PriorityArbitrator<DummyEnvironmentModel, DummyCommand>::Option::Flags;

    PriorityArbitrator<DummyEnvironmentModel, DummyCommand> testPriorityArbitrator;

    testPriorityArbitrator.addOption(testBehaviorHighPriority, OptionFlags::NO_FLAGS);
    testPriorityArbitrator.addOption(testBehaviorHighPriority, OptionFlags::NO_FLAGS);
    testPriorityArbitrator.addOption(testBehaviorMidPriority, OptionFlags::NO_FLAGS);
    testPriorityArbitrator.addOption(testBehaviorLowPriority, OptionFlags::NO_FLAGS);

    ASSERT_TRUE(testPriorityArbitrator.checkInvocationCondition(time, environmentModel));

    testPriorityArbitrator.gainControl(time, environmentModel);

    EXPECT_EQ("MidPriority", testPriorityArbitrator.getCommand(time, environmentModel));
    EXPECT_FALSE(testPriorityArbitrator.options().at(0)->verificationResult_.cached(time));
    EXPECT_FALSE(testPriorityArbitrator.options().at(1)->verificationResult_.cached(time));
    ASSERT_TRUE(testPriorityArbitrator.options().at(2)->verificationResult_.cached(time));
    // LowPriority could have been verified or not, so don't test it here

    EXPECT_TRUE(testPriorityArbitrator.options().at(2)->verificationResult_.cached(time).value()->isOk());
}


// Now DummyVerifier classifies the "MidPriority" command as invalid
TEST_F(CommandVerificationTest, DummyVerifierInPriorityArbitrator) {
    using OptionFlags = PriorityArbitrator<DummyEnvironmentModel, DummyCommand>::Option::Flags;

    PriorityArbitrator<DummyEnvironmentModel, DummyCommand> testPriorityArbitrator("PriorityArbitrator",
                                                                                   std::make_shared<DummyVerifier>());

    testPriorityArbitrator.addOption(testBehaviorHighPriority, OptionFlags::NO_FLAGS);
    testPriorityArbitrator.addOption(testBehaviorHighPriority, OptionFlags::NO_FLAGS);
    testPriorityArbitrator.addOption(testBehaviorMidPriority, OptionFlags::NO_FLAGS);
    testPriorityArbitrator.addOption(testBehaviorLowPriority, OptionFlags::NO_FLAGS);

    ASSERT_TRUE(testPriorityArbitrator.checkInvocationCondition(time, environmentModel));

    testPriorityArbitrator.gainControl(time, environmentModel);

    EXPECT_EQ("LowPriority", testPriorityArbitrator.getCommand(time, environmentModel));
    EXPECT_FALSE(testPriorityArbitrator.options().at(0)->verificationResult_.cached(time));
    EXPECT_FALSE(testPriorityArbitrator.options().at(1)->verificationResult_.cached(time));
    ASSERT_TRUE(testPriorityArbitrator.options().at(2)->verificationResult_.cached(time));
    ASSERT_TRUE(testPriorityArbitrator.options().at(3)->verificationResult_.cached(time));

    EXPECT_FALSE(testPriorityArbitrator.options().at(2)->verificationResult_.cached(time).value()->isOk());
    EXPECT_TRUE(testPriorityArbitrator.options().at(3)->verificationResult_.cached(time).value()->isOk());

    std::cout << "verificationResult for " << testPriorityArbitrator.options().at(2)->behavior_->name() << ": "
              << testPriorityArbitrator.options().at(2)->verificationResult_.cached(time).value() << std::endl;
    std::cout << "verificationResult for " << testPriorityArbitrator.options().at(3)->behavior_->name() << ": "
              << testPriorityArbitrator.options().at(3)->verificationResult_.cached(time).value() << std::endl;

    // clang-format off
    std::string expected_printout = invocationTrueString + commitmentTrueString + "PriorityArbitrator\n"
                        "    1. " + invocationFalseString + commitmentFalseString + "HighPriority\n"
                        "    2. " + invocationFalseString + commitmentFalseString + "HighPriority\n"
                        "    3. " + strikeThroughOn
                                  + invocationTrueString + commitmentFalseString + "MidPriority"
                                  + strikeThroughOff + "\n"
                        " -> 4. " + invocationTrueString + commitmentTrueString + "LowPriority";
    // clang-format on
    std::string actual_printout = testPriorityArbitrator.to_str(time, environmentModel);
    std::cout << actual_printout << std::endl;

    EXPECT_EQ(expected_printout, actual_printout);


    YAML::Node yaml = testPriorityArbitrator.toYaml(time, environmentModel);
    EXPECT_FALSE(yaml["options"][0]["verificationResult"].IsDefined());
    EXPECT_FALSE(yaml["options"][1]["verificationResult"].IsDefined());
    ASSERT_TRUE(yaml["options"][2]["verificationResult"].IsDefined());
    ASSERT_TRUE(yaml["options"][3]["verificationResult"].IsDefined());

    EXPECT_EQ("failed", yaml["options"][2]["verificationResult"].as<std::string>());
    EXPECT_EQ("passed", yaml["options"][3]["verificationResult"].as<std::string>());


    testPriorityArbitrator.loseControl(time, environmentModel);

    testBehaviorLowPriority->invocationCondition_ = false;
    ASSERT_TRUE(testPriorityArbitrator.checkInvocationCondition(time, environmentModel));

    testPriorityArbitrator.gainControl(time, environmentModel);

    EXPECT_THROW(testPriorityArbitrator.getCommand(time, environmentModel), NoApplicableOptionPassedVerificationError);
}


// Two "MidPriority" options, the second one defined as fallback option, so it should be selected, even if it is invalid
TEST_F(CommandVerificationTest, DummyVerifierInPriorityArbitratorWithFallback) {
    using OptionFlags = PriorityArbitrator<DummyEnvironmentModel, DummyCommand>::Option::Flags;

    PriorityArbitrator<DummyEnvironmentModel, DummyCommand> testPriorityArbitrator("PriorityArbitrator",
                                                                                   std::make_shared<DummyVerifier>());

    testPriorityArbitrator.addOption(testBehaviorHighPriority, OptionFlags::NO_FLAGS);
    testPriorityArbitrator.addOption(testBehaviorHighPriority, OptionFlags::NO_FLAGS);
    testPriorityArbitrator.addOption(testBehaviorMidPriority, OptionFlags::NO_FLAGS);
    testPriorityArbitrator.addOption(testBehaviorMidPriority, OptionFlags::FALLBACK);
    testPriorityArbitrator.addOption(testBehaviorLowPriority, OptionFlags::NO_FLAGS);

    ASSERT_TRUE(testPriorityArbitrator.checkInvocationCondition(time, environmentModel));

    testPriorityArbitrator.gainControl(time, environmentModel);

    EXPECT_EQ("MidPriority", testPriorityArbitrator.getCommand(time, environmentModel));
    EXPECT_FALSE(testPriorityArbitrator.options().at(0)->verificationResult_.cached(time));
    EXPECT_FALSE(testPriorityArbitrator.options().at(1)->verificationResult_.cached(time));
    ASSERT_TRUE(testPriorityArbitrator.options().at(2)->verificationResult_.cached(time));
    ASSERT_TRUE(testPriorityArbitrator.options().at(3)->verificationResult_.cached(time));
    EXPECT_FALSE(testPriorityArbitrator.options().at(4)->verificationResult_.cached(time));

    EXPECT_FALSE(testPriorityArbitrator.options().at(2)->verificationResult_.cached(time).value()->isOk());
    EXPECT_FALSE(testPriorityArbitrator.options().at(3)->verificationResult_.cached(time).value()->isOk());

    std::cout << "verificationResult for " << testPriorityArbitrator.options().at(2)->behavior_->name() << ": "
              << testPriorityArbitrator.options().at(2)->verificationResult_.cached(time).value() << std::endl;
    std::cout << "verificationResult for " << testPriorityArbitrator.options().at(3)->behavior_->name() << ": "
              << testPriorityArbitrator.options().at(3)->verificationResult_.cached(time).value() << std::endl;

    // clang-format off
    std::string expected_printout = invocationTrueString + commitmentTrueString + "PriorityArbitrator\n"
                        "    1. " + invocationFalseString + commitmentFalseString + "HighPriority\n"
                        "    2. " + invocationFalseString + commitmentFalseString + "HighPriority\n"
                        "    3. " + strikeThroughOn
                                  + invocationTrueString + commitmentFalseString + "MidPriority"
                                  + strikeThroughOff + "\n"
                        " -> 4. " + strikeThroughOn
                                  + invocationTrueString + commitmentFalseString + "MidPriority"
                                  + strikeThroughOff + "\n"
                        "    5. " + invocationTrueString + commitmentTrueString + "LowPriority";
    // clang-format on
    std::string actual_printout = testPriorityArbitrator.to_str(time, environmentModel);
    std::cout << actual_printout << std::endl;

    EXPECT_EQ(expected_printout, actual_printout);


    YAML::Node yaml = testPriorityArbitrator.toYaml(time, environmentModel);
    EXPECT_FALSE(yaml["options"][0]["verificationResult"].IsDefined());
    EXPECT_FALSE(yaml["options"][1]["verificationResult"].IsDefined());
    ASSERT_TRUE(yaml["options"][2]["verificationResult"].IsDefined());
    ASSERT_TRUE(yaml["options"][3]["verificationResult"].IsDefined());
    EXPECT_FALSE(yaml["options"][4]["verificationResult"].IsDefined());

    EXPECT_EQ("failed", yaml["options"][2]["verificationResult"].as<std::string>());
    EXPECT_EQ("failed", yaml["options"][3]["verificationResult"].as<std::string>());
}


TEST_F(CommandVerificationTest, DummyVerifierInCostArbitrator) {
    using OptionFlags = CostArbitrator<DummyEnvironmentModel, DummyCommand>::Option::Flags;

    CostArbitrator<DummyEnvironmentModel, DummyCommand> testCostArbitrator("CostArbitrator",
                                                                           std::make_shared<DummyVerifier>());

    CostEstimatorFromCostMap::CostMap costMap{{"HighPriority", 0}, {"MidPriority", 0.5}, {"LowPriority", 1}};
    CostEstimatorFromCostMap::Ptr costEstimator = std::make_shared<CostEstimatorFromCostMap>(costMap);

    testCostArbitrator.addOption(testBehaviorHighPriority, OptionFlags::NO_FLAGS, costEstimator);
    testCostArbitrator.addOption(testBehaviorHighPriority, OptionFlags::NO_FLAGS, costEstimator);
    testCostArbitrator.addOption(testBehaviorMidPriority, OptionFlags::NO_FLAGS, costEstimator);
    testCostArbitrator.addOption(testBehaviorLowPriority, OptionFlags::NO_FLAGS, costEstimator);

    ASSERT_TRUE(testCostArbitrator.checkInvocationCondition(time, environmentModel));

    testCostArbitrator.gainControl(time, environmentModel);

    EXPECT_EQ("LowPriority", testCostArbitrator.getCommand(time, environmentModel));
    EXPECT_FALSE(testCostArbitrator.options().at(0)->verificationResult_.cached(time));
    EXPECT_FALSE(testCostArbitrator.options().at(1)->verificationResult_.cached(time));
    ASSERT_TRUE(testCostArbitrator.options().at(2)->verificationResult_.cached(time));
    ASSERT_TRUE(testCostArbitrator.options().at(3)->verificationResult_.cached(time));

    EXPECT_FALSE(testCostArbitrator.options().at(2)->verificationResult_.cached(time).value()->isOk());
    EXPECT_TRUE(testCostArbitrator.options().at(3)->verificationResult_.cached(time).value()->isOk());

    // clang-format off
    std::string expectedPrintout = invocationTrueString + commitmentTrueString + "CostArbitrator\n"
                        "    - (cost:  n.a.) " + invocationFalseString + commitmentFalseString + "HighPriority\n"
                        "    - (cost:  n.a.) " + invocationFalseString + commitmentFalseString + "HighPriority\n"
                        "    - (cost:  n.a.) " + strikeThroughOn
                                  + invocationTrueString + commitmentFalseString + "MidPriority"
                                  + strikeThroughOff + "\n"
                        " -> - (cost: 1.000) " + invocationTrueString + commitmentTrueString + "LowPriority";
    // clang-format on
    std::string actualPrintout = testCostArbitrator.to_str(time, environmentModel);
    std::cout << actualPrintout << std::endl;

    EXPECT_EQ(expectedPrintout, actualPrintout);


    testCostArbitrator.loseControl(time, environmentModel);

    testBehaviorLowPriority->invocationCondition_ = false;
    ASSERT_TRUE(testCostArbitrator.checkInvocationCondition(time, environmentModel));

    testCostArbitrator.gainControl(time, environmentModel);

    EXPECT_THROW(testCostArbitrator.getCommand(time, environmentModel), NoApplicableOptionPassedVerificationError);
}

File: test/verification.py
import os
import time
import unittest
import yaml

import arbitration_graphs_py_with_verifier as ag_verifier


class TestCommandVerification(unittest.TestCase):
    def setUp(self):
        self.test_behavior_high_priority = ag_verifier.testing_types.DummyBehavior(
            False, False, "HighPriority"
        )
        self.test_behavior_mid_priority = ag_verifier.testing_types.DummyBehavior(
            True, False, "MidPriority"
        )
        self.test_behavior_low_priority = ag_verifier.testing_types.DummyBehavior(
            True, True, "LowPriority"
        )

        self.time = time.time()

    def test_dummy_verifier_in_priority_arbitrator(self):
        test_priority_arbitrator = ag_verifier.PriorityArbitrator(
            "PriorityArbitrator", ag_verifier.testing_types.DummyVerifier("MidPriority")
        )

        test_priority_arbitrator.add_option(
            self.test_behavior_high_priority,
            ag_verifier.PriorityArbitrator.Option.Flags.NO_FLAGS,
        )
        test_priority_arbitrator.add_option(
            self.test_behavior_high_priority,
            ag_verifier.PriorityArbitrator.Option.Flags.NO_FLAGS,
        )
        test_priority_arbitrator.add_option(
            self.test_behavior_mid_priority,
            ag_verifier.PriorityArbitrator.Option.Flags.NO_FLAGS,
        )
        test_priority_arbitrator.add_option(
            self.test_behavior_low_priority,
            ag_verifier.PriorityArbitrator.Option.Flags.NO_FLAGS,
        )

        self.assertTrue(test_priority_arbitrator.check_invocation_condition(self.time))

        test_priority_arbitrator.gain_control(self.time)

        self.assertEqual("LowPriority", test_priority_arbitrator.get_command(self.time))
        self.assertFalse(
            test_priority_arbitrator.options()[0].verification_result.cached(self.time)
        )
        self.assertFalse(
            test_priority_arbitrator.options()[1].verification_result.cached(self.time)
        )
        self.assertTrue(
            test_priority_arbitrator.options()[2].verification_result.cached(self.time)
        )
        self.assertTrue(
            test_priority_arbitrator.options()[3].verification_result.cached(self.time)
        )

        self.assertFalse(
            test_priority_arbitrator.options()[2]
            .verification_result.cached(self.time)
            .is_ok()
        )
        self.assertTrue(
            test_priority_arbitrator.options()[3]
            .verification_result.cached(self.time)
            .is_ok()
        )

        # Print verification results
        print(
            f"Verification result for {test_priority_arbitrator.options()[2].behavior.name}: "
            f"{test_priority_arbitrator.options()[2].verification_result.cached(self.time)}"
        )
        print(
            f"Verification result for {test_priority_arbitrator.options()[3].behavior.name}: "
            f"{test_priority_arbitrator.options()[3].verification_result.cached(self.time)}"
        )

        # fmt:off
        ps = ag_verifier.testing_types.print_strings
        expected_printout = (
            ps.invocation_true + ps.commitment_true + "PriorityArbitrator\n"
            "    1. " + ps.invocation_false + ps.commitment_false + "HighPriority\n"
            "    2. " + ps.invocation_false + ps.commitment_false + "HighPriority\n"
            "    3. " + ps.strike_through_on
                      + ps.invocation_true + ps.commitment_false + "MidPriority"
                      + ps.strike_through_off + "\n"
            " -> 4. " + ps.invocation_true + ps.commitment_true + "LowPriority"
        )
        # fmt:on
        actual_printout = test_priority_arbitrator.to_str(self.time)
        print(actual_printout)

        self.assertEqual(expected_printout, actual_printout)

        yaml_node = yaml.safe_load(test_priority_arbitrator.to_yaml_as_str(self.time))
        self.assertFalse("verificationResult" in yaml_node["options"][0])
        self.assertFalse("verificationResult" in yaml_node["options"][1])
        self.assertTrue("verificationResult" in yaml_node["options"][2])
        self.assertTrue("verificationResult" in yaml_node["options"][3])

        self.assertEqual("failed", yaml_node["options"][2]["verificationResult"])
        self.assertEqual("passed", yaml_node["options"][3]["verificationResult"])

        test_priority_arbitrator.lose_control(self.time)

        self.test_behavior_low_priority.invocation_condition = False
        self.assertTrue(test_priority_arbitrator.check_invocation_condition(self.time))

        test_priority_arbitrator.gain_control(self.time)

        with self.assertRaises(ag_verifier.NoApplicableOptionPassedVerificationError):
            test_priority_arbitrator.get_command(self.time)

    def test_dummy_verifier_in_priority_arbitrator_with_fallback(self):
        NO_FLAGS = ag_verifier.PriorityArbitrator.Option.Flags.NO_FLAGS
        FALLBACK = ag_verifier.PriorityArbitrator.Option.Flags.FALLBACK

        test_priority_arbitrator = ag_verifier.PriorityArbitrator(
            "PriorityArbitrator", ag_verifier.testing_types.DummyVerifier("MidPriority")
        )

        test_priority_arbitrator.add_option(self.test_behavior_high_priority, NO_FLAGS)
        test_priority_arbitrator.add_option(self.test_behavior_high_priority, NO_FLAGS)
        test_priority_arbitrator.add_option(self.test_behavior_mid_priority, NO_FLAGS)
        test_priority_arbitrator.add_option(self.test_behavior_mid_priority, FALLBACK)
        test_priority_arbitrator.add_option(self.test_behavior_low_priority, NO_FLAGS)

        self.assertTrue(test_priority_arbitrator.check_invocation_condition(self.time))

        test_priority_arbitrator.gain_control(self.time)

        self.assertEqual("MidPriority", test_priority_arbitrator.get_command(self.time))
        self.assertFalse(
            test_priority_arbitrator.options()[0].verification_result.cached(self.time)
        )
        self.assertFalse(
            test_priority_arbitrator.options()[1].verification_result.cached(self.time)
        )
        self.assertTrue(
            test_priority_arbitrator.options()[2].verification_result.cached(self.time)
        )
        self.assertTrue(
            test_priority_arbitrator.options()[3].verification_result.cached(self.time)
        )
        self.assertFalse(
            test_priority_arbitrator.options()[4].verification_result.cached(self.time)
        )

        self.assertFalse(
            test_priority_arbitrator.options()[2]
            .verification_result.cached(self.time)
            .is_ok()
        )
        self.assertFalse(
            test_priority_arbitrator.options()[3]
            .verification_result.cached(self.time)
            .is_ok()
        )

        print(
            f"verificationResult for {test_priority_arbitrator.options()[2].behavior.name}: "
            f"{test_priority_arbitrator.options()[2].verification_result.cached(self.time)}"
        )
        print(
            f"verificationResult for {test_priority_arbitrator.options()[3].behavior.name}: "
            f"{test_priority_arbitrator.options()[3].verification_result.cached(self.time)}"
        )

        # fmt:off
        ps = ag_verifier.testing_types.print_strings
        expected_printout = (
            ps.invocation_true + ps.commitment_true + "PriorityArbitrator\n"
            "    1. " + ps.invocation_false + ps.commitment_false + "HighPriority\n"
            "    2. " + ps.invocation_false + ps.commitment_false + "HighPriority\n"
            "    3. " + ps.strike_through_on
                      + ps.invocation_true + ps.commitment_false + "MidPriority"
                      + ps.strike_through_off + "\n"
            " -> 4. " + ps.strike_through_on
                      + ps.invocation_true + ps.commitment_false + "MidPriority"
                      + ps.strike_through_off + "\n"
            "    5. " + ps.invocation_true + ps.commitment_true + "LowPriority"
        )
        # fmt:on
        actual_printout = test_priority_arbitrator.to_str(self.time)
        print(actual_printout)

        self.assertEqual(expected_printout, actual_printout)

        yaml_node = yaml.safe_load(test_priority_arbitrator.to_yaml_as_str(self.time))
        self.assertFalse("verificationResult" in yaml_node["options"][0])
        self.assertFalse("verificationResult" in yaml_node["options"][1])
        self.assertTrue("verificationResult" in yaml_node["options"][2])
        self.assertTrue("verificationResult" in yaml_node["options"][3])
        self.assertFalse("verificationResult" in yaml_node["options"][4])

        self.assertEqual("failed", yaml_node["options"][2]["verificationResult"])
        self.assertEqual("failed", yaml_node["options"][3]["verificationResult"])

    def test_dummy_verifier_in_cost_arbitrator(self):
        NO_FLAGS = ag_verifier.CostArbitrator.Option.Flags.NO_FLAGS

        test_cost_arbitrator = ag_verifier.CostArbitrator(
            "CostArbitrator", ag_verifier.testing_types.DummyVerifier("MidPriority")
        )

        cost_map = {"HighPriority": 0, "MidPriority": 0.5, "LowPriority": 1}
        cost_estimator = ag_verifier.testing_types.CostEstimatorFromCostMap(cost_map)

        test_cost_arbitrator.add_option(
            self.test_behavior_high_priority, NO_FLAGS, cost_estimator
        )
        test_cost_arbitrator.add_option(
            self.test_behavior_high_priority, NO_FLAGS, cost_estimator
        )
        test_cost_arbitrator.add_option(
            self.test_behavior_mid_priority, NO_FLAGS, cost_estimator
        )
        test_cost_arbitrator.add_option(
            self.test_behavior_low_priority, NO_FLAGS, cost_estimator
        )

        self.assertTrue(test_cost_arbitrator.check_invocation_condition(self.time))

        test_cost_arbitrator.gain_control(self.time)

        self.assertEqual("LowPriority", test_cost_arbitrator.get_command(self.time))
        self.assertFalse(
            test_cost_arbitrator.options()[0].verification_result.cached(self.time)
        )
        self.assertFalse(
            test_cost_arbitrator.options()[1].verification_result.cached(self.time)
        )
        self.assertTrue(
            test_cost_arbitrator.options()[2].verification_result.cached(self.time)
        )
        self.assertTrue(
            test_cost_arbitrator.options()[3].verification_result.cached(self.time)
        )

        self.assertFalse(
            test_cost_arbitrator.options()[2]
            .verification_result.cached(self.time)
            .is_ok()
        )
        self.assertTrue(
            test_cost_arbitrator.options()[3]
            .verification_result.cached(self.time)
            .is_ok()
        )

        # fmt:off
        ps = ag_verifier.testing_types.print_strings
        expected_printout = (
            ps.invocation_true + ps.commitment_true + "CostArbitrator\n"
            "    - (cost:  n.a.) " + ps.invocation_false + ps.commitment_false + "HighPriority\n"
            "    - (cost:  n.a.) " + ps.invocation_false + ps.commitment_false + "HighPriority\n"
            "    - (cost: 0.500) " + ps.strike_through_on
                                   + ps.invocation_true + ps.commitment_false + "MidPriority"
                                   + ps.strike_through_off + "\n"
            " -> - (cost: 1.000) " + ps.invocation_true + ps.commitment_true + "LowPriority"
        )
        # fmt:on
        actual_printout = test_cost_arbitrator.to_str(self.time)
        print(actual_printout)

        self.assertEqual(expected_printout, actual_printout)

        test_cost_arbitrator.lose_control(self.time)

        self.test_behavior_low_priority.invocation_condition = False
        self.assertTrue(test_cost_arbitrator.check_invocation_condition(self.time))

        test_cost_arbitrator.gain_control(self.time)

        with self.assertRaises(ag_verifier.NoApplicableOptionPassedVerificationError):
            test_cost_arbitrator.get_command(self.time)


class TestCommandVerificationInConjunctiveCoordinator(unittest.TestCase):
    def setUp(self):
        self.test_behavior_a = ag_verifier.testing_types.DummyBehavior(
            False, False, "A"
        )
        self.test_behavior_b1 = ag_verifier.testing_types.DummyBehavior(
            True, False, "B"
        )
        self.test_behavior_c = ag_verifier.testing_types.DummyBehavior(True, True, "C")
        self.test_behavior_b2 = ag_verifier.testing_types.DummyBehavior(
            True, False, "B"
        )

        self.time = time.time()

    def test_verification(self):

        NO_FLAGS = ag_verifier.ConjunctiveCoordinator.Option.Flags.NO_FLAGS

        verifying_conjunctive_coordinator = ag_verifier.ConjunctiveCoordinator(
            "ConjunctiveCoordinator", ag_verifier.testing_types.DummyVerifier("B")
        )

        verifying_conjunctive_coordinator.add_option(self.test_behavior_a, NO_FLAGS)
        verifying_conjunctive_coordinator.add_option(self.test_behavior_b1, NO_FLAGS)
        verifying_conjunctive_coordinator.add_option(self.test_behavior_c, NO_FLAGS)
        verifying_conjunctive_coordinator.add_option(self.test_behavior_b2, NO_FLAGS)

        self.test_behavior_a.invocation_condition = True

        self.assertTrue(
            verifying_conjunctive_coordinator.check_invocation_condition(self.time)
        )
        self.assertFalse(
            verifying_conjunctive_coordinator.check_commitment_condition(self.time)
        )

        verifying_conjunctive_coordinator.gain_control(self.time)

        with self.assertRaises(ag_verifier.ApplicableOptionFailedVerificationError):
            verifying_conjunctive_coordinator.get_command(self.time)


if __name__ == "__main__":
    header = "Running " + os.path.basename(__file__)

    print("=" * len(header))
    print(header)
    print("=" * len(header) + "\n")
    unittest.main(exit=False)
    print("=" * len(header) + "\n")
